schema.go

```diff
@@ -71,3 +78,249 @@ func (schema DatabaseSchema) validateOperations(operations ...Operation) bool {
+	case TypeEnum:
+		typeStr = fmt.Sprintf("enum (type: %s): %v", column.TypeObj.Key.Type, column.TypeObj.Key.Enum)
+	case TypeMap:
+		typeStr = fmt.Sprintf("[%s]%s", column.TypeObj.Key.Type, column.TypeObj.Value.Type)
+	case TypeSet:
+		var keyStr string
+		if column.TypeObj.Key.Type == TypeUUID {
+			keyStr = fmt.Sprintf(" [%s (%s)]", column.TypeObj.Key.RefTable, column.TypeObj.Key.RefType)
+		} else {
+			keyStr = string(column.TypeObj.Key.Type)
+		}
+		typeStr = fmt.Sprintf("[]%s (min: %d, max: %d)", keyStr, column.TypeObj.Min, column.TypeObj.Max)
+	default:
+		typeStr = fmt.Sprintf("Unknown Type")
+	}
+
+	return fmt.Sprintf(strings.Join([]string{typeStr, flagStr}, " "))
+}
+
+// UnmarshalJSON unmarshalls a json-formatted column
+func (column *ColumnSchema) UnmarshalJSON(data []byte) error {
+	var _col ColumnJSON
+	// Unmarshall known keys
+	if err := json.Unmarshal(data, &_col); err != nil {
+		return err
+	}
+
+	column.Name = _col.Name
+	column.Ephemeral = _col.Ephemeral
+	column.Mutable = _col.Mutable
+
+	// 'type' migth be a string or an object, let's figure it out
+	var typeString string
+	if err := json.Unmarshal(_col.TypeRawMsg, &typeString); err == nil {
+		// This was an easy one. Use the string as our 'extended' type
+		column.Type = ExtendedType(typeString)
+		return nil
+	}
+
+	// type can be an object defined as:
+	// "key": <base-type>                 required
+	// "value": <base-type>               optional
+	// "min": <integer>                   optional
+	// "max": <integer> or "unlimited"    optional
+	column.TypeObj = &ColumnTypeObject{
+		Key:   &BaseType{},
+		Value: nil,
+		Max:   1,
+		Min:   1,
+	}
+	if err := json.Unmarshal(_col.TypeRawMsg, column.TypeObj); err != nil {
+		return err
+	}
+
+	// Now we have to unmarshall some fields manually because they can store
+	// values of different types. Also, in order to make really know what
+	// native type can store a value of this column, the RFC defines some logic
+	// based on the values of 'type'. So, in addition to manually unmarshalling,
+	// let figure out what is the real native type and store it in column.Type
+	// for ease of use.
+
+	// 'max' can be an integer or the string "unlimmited". To simplify, use "-1"
+	// as unlimited
+	if column.TypeObj.MaxRawMsg != nil {
+		var maxString string
+		if err := json.Unmarshal(*column.TypeObj.MaxRawMsg, &maxString); err == nil {
+			if maxString == "unlimited" {
+				column.TypeObj.Max = Unlimited
+			} else {
+				return fmt.Errorf("Unknown max value %s", maxString)
+			}
+		} else if err := json.Unmarshal(*column.TypeObj.MaxRawMsg, &column.TypeObj.Max); err != nil {
+			return err
+		}
+	}
+
+	// 'key' and 'value' can, themselves, be a string or a BaseType.
+	// key='<atomic_type>' is equivalent to 'key': {'type': '<atomic_type>'}
+	// To simplify things a bit, we'll translate the former to the latter
+	if err := json.Unmarshal(*column.TypeObj.KeyRawMsg, &column.TypeObj.Key.Type); err != nil {
```

Comment:
can be detect such stuff without double unmarshal?

Reply:
How? maybe you can unmarshal it to a `interface{}` and then infer it's type using reflex, but what's the difference between that and storing the `RawMessage` and perform conditional unmarshalling?

Reply:
@vtolstov, I am pretty sure of what to do with regards to the rest of the comments (thanks, by the way :))

For this one, I have my doubts. I have pushed an additional commit on top of the entire PR modifying the implementation to only call unmarshal once.

PTAL at it https://github.com/eBay/libovsdb/pull/30/commits/e329dbbe7306ff6b78b16db34e164b7bf244c15f

I think I would prefer to stick with the double-unmarshal implementation, though it may still be place for improvement, I'm sure.
WDYT? Do you have any other suggestion?

Reply:
For the record, after discussing with @vtolstov on slack and profiling several options, we have implemented this using multiple rounds of json.Unmarshal but making temporary json objects private and ... well ... really temporary so GC cleans it all afterwards.